js/client-battle.js

```diff
@@ -1150,6 +1200,231 @@
 		},
 		hideTooltip: function () {
 			$('#tooltipwrapper').html('');
+		},
+		// Functions to calculate speed ranges of an opponent.
+		getTemplateMinSpeed: function (template, level) {
+			var nature = (this.battle.gen < 3 || this.battle.tier === 'Random Battle') ? 1 : 0.9;
+			return Math.floor(Math.floor(2 * template.baseStats['spe'] * level / 100 + 5) * nature);
+		},
+		getTemplateMaxSpeed: function (template, level) {
+			var nature = (this.battle.gen < 3 || this.battle.tier === 'Random Battle') ? 1 : 1.1;
+			var maxIvs = (this.battle.gen > 2) ? 31 : 30;
+			return Math.floor(Math.floor(Math.floor(2 * template.baseStats['spe'] + maxIvs + 63) * level / 100 + 5) * nature);
+		},
+		// Gets the proper current type for moves with a variable type.
+		getMoveType: function(move, pokemon) {
+			var moveType = move.type;
+			// Normalize is the first move type changing effect.
+			if (pokemon.ability === 'Normalize') {
+				moveType = 'Normal';
+			}
+			// Moves that require an item to change their type.
+			if (!this.battle.hasPseudoWeather('Magic Room') && !pokemon.volatiles['embargo']) {
+				if (move.id === 'judgment') {
+					var item = Tools.getItem(pokemon.item);
+					if (item.onPlate) moveType = item.onPlate;
+				}
+				if (move.id === 'technoblast') {
+					var item = Tools.getItem(pokemon.item);
+					if (item.onDrive) moveType = item.onDrive;
+				}
+				if (move.id === 'naturalgift') {
+					var item = Tools.getItem(pokemon.item);
+					if (item.naturalGift) moveType = item.naturalGift.type;
+				}
+			}
+			// Weather and pseudo-weather type changes.
+			if (move.id === 'weatherball') {
+				if (this.battle.weather === 'sunnyday' || this.battle.weather === 'desolateland') moveType = 'Fire';
+				if (this.battle.weather === 'raindance' || this.battle.weather === 'primordialsea') moveType = 'Water';
+				if (this.battle.weather === 'sandstorm') moveType = 'Rock';
+				if (this.battle.weather === 'hail') moveType = 'Ice';
```

Comment:
It's not possible to know whether Cloud Nine or Air Lock are actually active, since it could be a hackmons format or something.